### PR TITLE
GenAI: define conventions for embeddings operations

### PR DESCRIPTION
**Pull request by trentm**
Refs: #1174

---

Many LLMs support an Embeddings API, for example:

- [OpenAI](https://platform.openai.com/docs/api-reference/embeddings)
- [Cohere](https://docs.cohere.com/reference/embed)
- [Voyage AI](https://docs.anthropic.com/en/docs/build-with-claude/embeddings#voyage-http-api)
- [Amazon Bedrock Titan](https://docs.aws.amazon.com/bedrock/latest/userguide/model-parameters-titan-embed-text.html)

This proposal defines OpenTelemetry semantic conventions to use for instrumenting Embeddings API client usage.

# Overview

A *span* will be created for Embeddings API calls. The only differences from existing chat spans are:

- A new `embeddings` value is added for attribute `gen_ai.operation.name`.
- A new `gen_ai.request.encoding_formats` attribute is defined. It is only relevant for embeddings operations.
- Many of the "recommended" span attributes do not apply to embeddings operations. Perhaps that is fine as currently defined. (**Open question:** should a `note: ...` be added or `if applicable.` added to the `brief` for those span attributes?)

No new *events* are proposed. See "Notes" below.

For *metrics*:

- The existing `gen_ai.client.token.usage` metric applies, with the note that *output* tokens do not apply for embeddings operations, so the metric would only be recorded with the `gen_ai.token.type: 'input'` attribute (as already specified).
- The existing `gen_ai.client.operation.duration` metric applies as currently specified.

# Example span

This an example embeddings operation span using the `openai` client library and the OTel JS ConsoleSpanExporter:

```
{
  traceId: '1dea787b9cd7bb895aee5bb74090610d',
  parentId: undefined,
  traceState: undefined,
  name: 'embeddings text-embedding-ada-002',
  id: '65aafaf145cf535a',
  kind: 2,
  timestamp: 1732229356238000,
  duration: 551555.417,
  attributes: {
    'gen_ai.operation.name': 'embeddings',
    'gen_ai.request.model': 'text-embedding-ada-002',
    'gen_ai.system': 'openai',
    'gen_ai.request.encoding_formats': [ 'float' ],
    'gen_ai.response.model': 'text-embedding-ada-002',
    'gen_ai.usage.input_tokens': 9
  },
  status: { code: 0 },
  events: [],
  links: []
}
```

# Notes

This section contains supporting notes and reasoning for some of the proposed values.

- A span attribute for the `dimensions` parameter (in the OpenAI API) was considered, but dropped as likely not being useful. Happy to revisit that if others know of a reasonable use case.

- A (log) event to record the *input strings* to the Embeddings API call is **not** being proposed. The reasoning is that API calls for embeddings are expected to be higher volume, and the input strings less valuable to application observability than *chat* content, so the cost-benefit ratio is much less. If a good use case is presented for optionally recording Embeddings input strings, then this can be revisited.

- Recording Embeddings API response *vectors* in telemetry is **not** being proposed. Vectors are large and would not be useful for observability.

- The operation name `embeddings` was selected. Other possible options considered:

    1. `embeddings`
    2. `embedding`
    3. `embed`

    My inclination is `embeddings` to match the (OpenAI) API name: `embeddings.create`.
    [Cohere's API](https://docs.cohere.com/reference/embed) name is `embed`.
    LangTrace uses `embed`.
    Current semconv values are `chat` (e.g. for `openai.chat.completions.create()`) and `text_completion` (for the deprecated `openai.chat.create()`).

- The Cohere and Anthropic APIs have a request attribute `input_type`, for creating embeddings for inputs other than text. This might be worth considering adding as well. I have not currently proposed this because I have only prototyped with OpenAI.

### CODE REVIEW

File: model/gen-ai/registry.yaml
Comment (trentm) [inline model/gen-ai/registry.yaml:131]: Reviewer note: In the PR description I added an "Open question" about whether it would be good/helpful to add "If applicable" notes (or additions to the `brief`) for the span attributes that do not apply to embeddings operations. This is one such note to feel it out. Thoughts?

**Review by xrmx — APPROVED** [review]

File: model/gen-ai/registry.yaml
Comment (xrmx) [reply model/gen-ai/registry.yaml:131]: Does this "to do" should be "do not"?

File: model/gen-ai/registry.yaml
Comment (xrmx) [inline model/gen-ai/registry.yaml:101]: Maybe these `APIs` should be `GenAI systems` to match the attribute?

Comment (trentm) [reply model/gen-ai/registry.yaml]: Thanks. commit 2350ab0

File: docs/attributes-registry/gen-ai.md
Comment (xrmx) [inline docs/attributes-registry/gen-ai.md:51]:
```suggestion
**[4] `gen_ai.usage.output_tokens`:** If applicable. For example, embeddings operations do not use output tokens.
```

File: CONTRIBUTING.md
Comment (lmolkova) [reply CONTRIBUTING.md:117]: thanks, let's keep it!

File: model/gen-ai/registry.yaml
Comment (lmolkova) [reply model/gen-ai/registry.yaml:131]: "if applicable" usually goes on the `requirement_level" when referencing an attribute. No need to add it on the registry level - all attributes are used when they are appliable

**Review by lmolkova — APPROVED** [review]
Left a few minor comments, looks great otherwise!

File: model/gen-ai/registry.yaml
Comment (trentm) [reply model/gen-ai/registry.yaml]: @lmolkova 

1. Do you mean like this?

https://github.com/open-telemetry/semantic-conventions/blob/5517079850f4a1bedfbb376354e2a2dfe74edc86/model/process/registry.yaml#L107-L118

That's the only example I've seen of `requirement_level` with both some explanatory text and a key other than `conditionally_required`.

See commit d79bf33.

2. Do you think there is value adding this kind of "if applicable for the particular GenAI operation" to most of the attributes? Or does that end up being too much table node spam?

File: docs/attributes-registry/gen-ai.md
Comment (trentm) [reply docs/attributes-registry/gen-ai.md]: commit b053fa1. Thanks.

File: model/gen-ai/spans.yaml
Comment (trentm) [reply model/gen-ai/spans.yaml:28]: I would be fine with either option. I went with the latter in commit cdb9e09.

**Review by lmolkova — APPROVED** [review]
Looks great!

**Comment by trentm** [comment]
^^ the check failure is:

```
   ERROR: 1 dead links found in ./docs/system/system-metrics.md !
  [✖] https://blogs.oracle.com/linux/post/understanding-linux-kernel-memory-statistics → Status: 0
make: *** [Makefile:72: markdown-link-check] Error 1
```

I'm guessing it was a fluke server error response from that site. The URL exists for me.  The check passes for me locally as well:

```
% make markdown-link-check
semantic-conventions2@ /Users/trentm/tm/semantic-conventions2
└── markdown-link-check@3.11.2
```

I don't have permissions to re-run that workflow run.

**Comment by lmolkova** [comment]
the link check is not required and oracle blog is the usual offender. We need another approval to merge though. @open-telemetry/semconv-genai-approvers ptal!

**Review by karthikscale3 — APPROVED** [review]
Thanks for starting this

**Review by alizenhom — APPROVED** [review]

**Review by AlexanderWert — APPROVED** [review]

**Comment by axiomofjoy** [comment]
Just wanted to add some thoughts on observability workflows for embeddings since they are important to our product at Arize. We use both embedding vectors and their associated text for debugging and troubleshooting workflows. Because embedding vectors map semantically similar content to nearby vectors, you can use them for observability workflows such as:

- clustering to find semantically meaningful groups of user queries, and then measuring performance/ computing evaluation metrics on those clusters to gain granular understanding of your application performance on a meaningful subset of data
- embedding-based similarity search to find semantically similar examples to a chosen example, e.g., for debugging or curating datasets for fine-tuning
- understanding how query and corpus distributions differ (e.g., to identify out-of-distribution queries that likely do not have answers in the embedded corpus)

These workflows require both embedding vectors and associated content (text, image, etc.), which we include as semantic conventions in the OpenInference spec.

I've included a few resources below for additional context.

https://github.com/user-attachments/assets/1125b701-c04e-4672-b308-a54029860e88

- https://www.youtube.com/watch?v=K532ZClP-xQ
- https://www.youtube.com/watch?v=JaFt1U5rxs4

**Pull request merged**
